File: alle-core/src/main/java/io/github/shomah4a/alle/core/command/Command.java
package io.github.shomah4a.alle.core.command;

import java.util.concurrent.CompletableFuture;

/**
 * エディタの操作単位。
 * 名前を持ち、コンテキストを受け取って実行する。
 */
public interface Command {

    /**
     * コマンド名を返す。
     */
    String name();

    /**
     * コマンドを実行する。
     */
    CompletableFuture<Void> execute(CommandContext context);

    /**
     * コマンド実行後にリージョン（mark）を維持するかどうかを返す。
     * trueを返すコマンドの実行後はmarkがクリアされない。
     * 移動系コマンドやset-markコマンドがtrueを返す。
     */
    default boolean keepsRegionActive() {
        return false;
    }
}

File: alle-script/src/main/resources/io/github/shomah4a/alle/script/alle/internal/command.py
"""CommandBase から Java Command への変換ブリッジ。"""

from __future__ import annotations

from typing import Any

import java

from alle.command import CommandBase
from alle.context import CommandContext

ScriptCommand: Any = java.type('io.github.shomah4a.alle.script.ScriptCommand')


def make_command(base: CommandBase) -> Any:
    """CommandBase から Java ScriptCommand インスタンスを生成する。

    Java側のScriptCommandがCommandContextをScriptCommandContextに変換し、
    Python側でさらにCommandContextラッパーに包んでからrun関数を呼び出す。

    :param base: 変換元のコマンド
    :type base: CommandBase
    :return: Java Command インスタンス
    :rtype: Command
    """
    run_fn = base.run

    def wrapped_run(java_script_ctx: Any) -> None:
        ctx = CommandContext(java_script_ctx)
        run_fn(ctx)

    return ScriptCommand(base.name(), wrapped_run)
